lib/rest-api.php

```diff
@@ -637,3 +684,115 @@ function gutenberg_filter_user_query_arguments( $prepared_args, $request ) {
 	return $prepared_args;
 }
 add_filter( 'rest_user_query', 'gutenberg_filter_user_query_arguments', 10, 2 );
+
+/**
+ * Overload taxonomy and term permission handling to address our new necessary behavior.
+ *
+ * This is temporary code that will be removed once the Trac ticket lands in a release.
+ *
+ * @see https://core.trac.wordpress.org/ticket/44096
+ *
+ * @param WP_HTTP_Response $response Result to send to the client. Usually a WP_REST_Response.
+ * @param WP_REST_Server   $handler  ResponseHandler instance (usually WP_REST_Server).
+ * @param WP_REST_Request  $request  Request used to generate the response.
+ * @return $response
+ */
+function gutenberg_filter_request_after_callbacks( $response, $handler, $request ) {
+	$should_rerun_response = false;
+	if ( is_wp_error( $response ) ) {
+		// Handle GET /wp/v2/taxonomies?context=edit when user can assign_terms
+		// but not manage_terms.
+		if ( '/wp/v2/taxonomies' === $request->get_route()
+			&& is_array( $handler['permission_callback'] )
+			&& is_a( $handler['permission_callback'][0], 'WP_REST_Taxonomies_Controller' )
+			&& 'edit' === $request['context']
+			&& 'rest_cannot_view' === $response->get_error_code() ) {
+			if ( ! empty( $request['type'] ) ) {
+				$taxonomies = get_object_taxonomies( $request['type'], 'objects' );
+			} else {
+				$taxonomies = get_taxonomies( '', 'objects' );
+			}
+			foreach ( $taxonomies as $taxonomy ) {
+				if ( ! empty( $taxonomy->show_in_rest )
+					&& current_user_can( $taxonomy->cap->assign_terms ) ) {
+					$GLOBALS['Gutenberg_Temporary_Taxonomies_Controller'] = $handler['permission_callback'][0];
+
+					$handler['callback']   = 'gutenberg_taxonomies_controller_get_items';
+					$should_rerun_response = true;
+					break;
+				}
+			}
+		}
+		// Handle POST /wp/v2/tags (and non-hiearchical taxonomies) when user
+		// can assign_terms but not manage terms. Users should be able to create
+		// terms.
+		if ( 'rest_cannot_create' === $response->get_error_code()
+			&& is_array( $handler['permission_callback'] )
+			&& is_a( $handler['permission_callback'][0], 'WP_REST_Terms_Controller' ) ) {
+			// 'taxonomy' is a protected attribute, and PHP 5.2 doesn't support
+			// ReflectionProperty->setAccessible(), so we need to get really dirty.
```

Comment:
I'm not going to lie, I'm impressed with this hack.

But, can't you get the value of `::taxonomy` from the data returned by `::get_item_schema()`?

Reply:
> But, can't you get the value of `::taxonomy` from the data returned by `::get_item_schema()`?

You can, in fact. I'd forgotten about that. 0c9c3294347220707e73979ad2d3425b4b39c044